Allow oci-systemd hooks to properly engage

fbladilo opened the pull request:
- Removed ENTRYPOINT which conflicted with oci-systemd hooks
- Added container env variable for systemd use and volume entries for docker hints
- Tested on OCP 3.5

fbladilo committed:
Allow oci-systemd hooks to properly engage

- Removed ENTRYPOINT which conflicted with oci-systemd hooks
- Added container env variable for systemd use and volume entries for docker hints
- Tested on OCP 3.5

fbladilo commented:
@abellotti @carbonin @bdunne With these changes, the OCI systemd hooks are engaging properly and as far as I can tell the httpd pod can be run under "anyuid" instead of customized SCC. 

```
[root@franco-ocp-master container-httpd (fix_oci_systemd_hook)]# oc version
oc v3.5.5.5
kubernetes v1.5.2+43a9be4
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://franco-ocp-master.e2e.bos.redhat.com:8443
openshift v3.5.5.5
kubernetes v1.5.2+43a9be4
[root@franco-ocp-master container-httpd (fix_oci_systemd_hook)]# oc get pods
NAME                 READY     STATUS    RESTARTS   AGE
httpd-1-bfklv        1/1       Running   0          24m
manageiq-0           1/1       Running   1          24m
memcached-1-1wdr4    1/1       Running   0          24m
postgresql-1-z48dq   1/1       Running   0          24m

root@franco-ocp-master container-httpd (fix_oci_systemd_hook)]# oc describe pods httpd-1-bfklv | grep "Security Policy"
Security Policy:	anyuid
[root@franco-ocp-master container-httpd (fix_oci_systemd_hook)]# oc rsh httpd-1-bfklv 
sh-4.2# ps -ef
UID         PID   PPID  C STIME TTY          TIME CMD
root          1      0  0 19:21 ?        00:00:00 /usr/sbin/init
root         18      1  0 19:21 ?        00:00:00 /usr/lib/systemd/systemd-journ
root         19      1  0 19:21 ?        00:00:00 /usr/sbin/httpd -DFOREGROUND
dbus         55      1  0 19:21 ?        00:00:00 /bin/dbus-daemon --system --ad
apache       89     19  0 19:21 ?        00:00:00 /usr/sbin/httpd -DFOREGROUND
apache       90     19  0 19:21 ?        00:00:00 /usr/sbin/httpd -DFOREGROUND
apache       91     19  0 19:21 ?        00:00:00 /usr/sbin/httpd -DFOREGROUND
apache       92     19  0 19:21 ?        00:00:00 /usr/sbin/httpd -DFOREGROUND
apache       93     19  0 19:21 ?        00:00:00 /usr/sbin/httpd -DFOREGROUND
apache       94     19  0 19:26 ?        00:00:00 /usr/sbin/httpd -DFOREGROUND

```

Fryguy commented on Dockerfile:
```diff
@@ -58,15 +61,12 @@ RUN (cd /lib/systemd/system/sysinit.target.wants && for i in *; do [ $i == syste
 WORKDIR /etc/httpd
 
-RUN systemctl enable dbus httpd
+RUN systemctl enable httpd
```
But I thought we needed dbus for a feature. cc @abellotti

carbonin replied:
It's clearly running in @fbladilo 's comment. It's possible it's enabled by default and we just didn't need to run it again.

fbladilo replied:
@Fryguy Systemd brings up dbus , it's redundant.

carbonin commented:
So were the hooks somehow using the `entrypoint` to run?

Fryguy commented:
This is going to conflict with #12 .  Can you work with @abellotti to ensure that his changes and your changes are tested together?

Fryguy commented:
> as far as I can tell the httpd pod can be run under "anyuid" instead of customized SCC.

😍

abellotti commented on Dockerfile:
```diff
@@ -58,15 +61,12 @@ RUN (cd /lib/systemd/system/sysinit.target.wants && for i in *; do [ $i == syste
-VOLUME [ "/sys/fs/cgroup" ]
+VOLUME /tmp /run /var/log/httpd
 
-ENTRYPOINT [ "entrypoint" ]
```
I do update entrypoint to do some initialization needed based on the new auth-config @fbladilo as per https://github.com/ManageIQ/container-httpd/pull/12/files#diff-ea90b4f9a644419e1278f7b629a37007  what else can be done at startup that will keep OCI happy ?

fbladilo replied:
@abellotti I need to test, I think OCI hooks would be happy as long as ENTRYPOINT is not set to something else, we could potentially hijack the CMD, perform work and then exec the real /sbin/init. Another option is to create an initialization systemd service unit that performs the pre-work and is a dependency to bring up httpd. Like we use to have before with "evminit" and "appliance-initialize" units. Thoughts?

abellotti commented on Dockerfile:
```diff
@@ -1,6 +1,9 @@
 MAINTAINER ManageIQ https://github.com/ManageIQ/manageiq-appliance-build
 
+## Systemd
+ENV container oci
```
does this affect running with minishift ?

fbladilo replied:
@abellotti It should not , the variable is a hint to systemd that it is running inside a container.

fbladilo commented:
@carbonin They seem to be processing ENTRYPOINT, instead of CMD. Now this might be something particular to OpenShift will need to test standalone docker to know more. The resulting httpd POD was lacking proper systemd CGROUP setup and failing.

carbonin commented on Dockerfile:
```diff
@@ -58,15 +61,12 @@ RUN (cd /lib/systemd/system/sysinit.target.wants && for i in *; do [ $i == syste
-RUN systemctl enable dbus httpd
+RUN systemctl enable httpd
 
-VOLUME [ "/sys/fs/cgroup" ]
```
We still need this line to get things running in minishift, will adding it back cause problems in OCP?

fbladilo replied:
@carbonin I tested and it works with the extra cgroup on VOLUMES on OCP. I'm almost certain minishift does not use oci systemd hooks, that would also explain the need for miq-sysadmin SCC.

carbonin replied:
Okay, so can we add this volume back then we will doc in the pods repo that you can assign different sccs based on where you are running.

fbladilo committed:
Restore /sys/fs/cgroup volume to aid minishift deployments

- Removed /tmp and /run docker volume hints, oci-systemd-hook will tmpfs these when engaged
- Removed /var/log/httpd , this volume is supplied by the docker lvm layer

fbladilo commented:
One more change to help minishift deployments, summary of the current situation : 

OCP w/ hooks : 

Does not need much image volume assistance but conflicts when ENTRYPOINT is set to something other than "init" or "systemd". When engaged it will mount cgroup appropiately and also mount tmpfs /run and /tmp automatically. It will run with anyuid SCC.

```
[root@franco-ocp-master container-httpd (fix_oci_systemd_hook)]# oc describe pods httpd-3-ttw25 | grep "Security Policy"
Security Policy:	anyuid
[root@franco-ocp-master container-httpd (fix_oci_systemd_hook)]# oc rsh httpd-3-ttw25 
sh-4.2# mount | grep -e "/run " -e "/tmp" -e "system.slice"
tmpfs on /run type tmpfs (rw,nosuid,nodev,relatime,context="system_u:object_r:svirt_sandbox_file_t:s0:c2,c8",size=65536k,mode=755)
tmpfs on /tmp type tmpfs (rw,nosuid,nodev,relatime,context="system_u:object_r:svirt_sandbox_file_t:s0:c2,c8",size=4194304k)
cgroup on /sys/fs/cgroup/systemd/system.slice/docker-72ab3bbced7c5268ef6e34b81742b63e178b5d810fa7c101c0f6b4fc4c637120.scope type cgroup (rw,nosuid,nodev,noexec,relatime,xattr,release_agent=/usr/lib/systemd/systemd-cgroups-agent,name=systemd)
sh-4.2# pidof httpd
102 93 92 91 90 89 19
sh-4.2# 
[root@franco-ocp-master container-httpd (fix_oci_systemd_hook)]# oc version
oc v3.5.5.5
kubernetes v1.5.2+43a9be4
features: Basic-Auth GSSAPI Kerberos SPNEGO

[root@franco-ocp-master container-httpd (fix_oci_systemd_hook)]# rpm -q oci-systemd-hook
oci-systemd-hook-0.1.8-4.1.gite533efa.el7.x86_64
[root@franco-ocp-master container-httpd (fix_oci_systemd_hook)]# rpm -q container-selinux
container-selinux-2.21-1.el7.noarch
```
Minishift without hooks : 

Does not conflict with ENTRYPOINT, does need an extra VOLUME hint on image for cgroup fs as well as a special SCC (miq-sysadmin) in order to access host subsystems.

We opened an issue regarding this on their repo : 

https://github.com/minishift/minishift/issues/1299

Something to consider is that while minishift is a great devel tool it does not resemble openshift production clusters.

In the meantime, we are discussing some alternatives to ENTRYPOINT for Alberto's work on other related PRs.

carbonin commented on Dockerfile:
```diff
@@ -58,15 +61,12 @@ RUN (cd /lib/systemd/system/sysinit.target.wants && for i in *; do [ $i == syste
-VOLUME [ "/sys/fs/cgroup" ]
+VOLUME /sys/fs/cgroup
 
-ENTRYPOINT [ "entrypoint" ]
```
Can we remove the entrypoint script if it is no longer being used?

fbladilo replied:
👍

fbladilo committed:
Removed unused entrypoint

carbonin reviewed (approved):
I'm good with this as it stands. Will wait for a :+1: from @abellotti first to make sure hes set with the auth side of things re: entrypoint.

bdunne reviewed (approved):
👍 LGTM

carbonin commented:
@abellotti Just waiting on a :+1: from you on removing entrypoint to merge this.

abellotti commented:
👍 from me.

Auth-Conf PRs can be updated accordingly, with careful ConditionPathExists/Before directives for the new auth-config service to guarantee execution after the lifecycle postStart entrypoint hook and before networking starts.
